File: DemoServer/AdditionalLanguages/go/Queries/projectingIndividualFields/ProjectingIndividualFields.go
package main

//region Usings
import "github.com/ravendb/ravendb-go-client"
//endregion

var globalDocumentStore *ravendb.DocumentStore

func main() {
    createDocumentStore()
    createDatabase()
    projectingIndividualFieldsInQuery()
    globalDocumentStore.Close()
}

func createDocumentStore() (*ravendb.DocumentStore, error) {
    if globalDocumentStore != nil {
        return globalDocumentStore, nil
    }
    urls := []string{"http://localhost:8080"}
    store := ravendb.NewDocumentStore(urls, "testGO")
    err := store.Initialize()
    if err != nil {
        return nil, err
    }
    globalDocumentStore = store
    return globalDocumentStore, nil
}

func createDatabase() {
    databaseRecord := ravendb.NewDatabaseRecord()
    databaseRecord.DatabaseName = "testGO"
    createDatabaseOperation := ravendb.NewCreateDatabaseOperation(databaseRecord, 1)
    var err = globalDocumentStore.Maintenance().Server().Send(createDatabaseOperation)
    if err != nil {
        fmt.Printf("d.store.Maintenance().Server().Send(createDatabaseOperation) failed with %s\n", err)
    }
}

//region Demo
func projectingIndividualFieldsInQuery() error {

    session, err := globalDocumentStore.OpenSession("")
    if err != nil {
        return err
    }
    defer session.Close()

    //region Step_1
    queriedType := reflect.TypeOf(&Company{})
    projectedQuery := session.QueryCollectionForType(queriedType)
    //endregion
    
    //region Step_2
    projectedType := reflect.TypeOf(&CompanyDetails{})
    
    fields := []string{"Name", "Address.City", "Address.Country"}
    projections := []string{"CompanyName", "City", "Country"}
    
    queryData := &ravendb.QueryData{
        Fields:      fields,
        Projections: projections,
    }    
    projectedQuery = projectedQuery.SelectFieldsWithQueryData(projectedType, queryData)
    //endregion
    
    //region Step_3
    var projectedResults []*CompanyDetails
    err = projectedQuery.GetResults(&projectedResults)
    if err != nil {
        return err
    }
    //endregion
    
    return nil
}

type Company struct {
    ID      string
    Name    string
    Phone   string
    Contact *Contact
}
type Contact struct {
    Name  string
    Title string
}
type CompanyDetails struct {
    CompanyName string
    City        string
    Country     string
}
//endregion

File: DemoServer/AdditionalLanguages/java/src/main/java/net/ravendb/demo/queries/projectingIndividualFields/ProjectingIndividualFields.java
package net.ravendb.demo.queries.projectingIndividualFields;

//region Usings
import net.ravendb.client.documents.queries.QueryData;
import net.ravendb.client.documents.session.IDocumentQuery;
import net.ravendb.client.documents.session.IDocumentSession;
import java.util.List;
//endregion
import net.ravendb.demo.common.DocumentStoreHolder;
import net.ravendb.demo.common.models.Company;

public class ProjectingIndividualFields {
    //region Demo
    public static class CompanyDetails {
        private String companyName;
        private String city;
        private String country;
    //endregion
        public String getCompanyName() {
            return companyName;
        }

        public void setCompanyName(String companyName) {
            this.companyName = companyName;
        }

        public String getCity() {
            return city;
        }

        public void setCity(String city) {
            this.city = city;
        }

        public String getCountry() {
            return country;
        }

        public void setCountry(String country) {
            this.country = country;
        }
    //region Demo
    }
    //endregion

    public List<CompanyDetails> run() {
        List<CompanyDetails> projectedResults;

        //region Demo
        try (IDocumentSession session = DocumentStoreHolder.store.openSession()) {
            //region Step_1
            IDocumentQuery<CompanyDetails> projectedQuery = session.query(Company.class)
            //endregion
            //region Step_2
                .selectFields(CompanyDetails.class,
                    new QueryData(
                        new String[] { "Name", "Address.City", "Address.Country" },
                        new String[] { "CompanyName", "City", "Country" }));
            //endregion

            //region Step_3
            projectedResults = projectedQuery.toList();
            //endregion
        }
        //endregion
        return projectedResults;
    }
}
